### ISSUE TITLE
setuptools 51.3.3 breaks ruamel.yaml

### ISSUE THREAD

**Issue by asavah** [pypa/setuptools#2540]
Note: no other changes to my setup has been at this time except upgrading setuptools.

51.3.3 (bad):
```
python3
Python 3.9.1 (default, Dec 18 2020, 18:01:25) 
[GCC 10.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import ruamel.yaml
>>> dir(ruamel.yaml)
['__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', 'pkg_resources']
>>> dir(ruamel.yaml.safe_load)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'ruamel.yaml' has no attribute 'safe_load'
```

51.1.2 (good):
```
python3
Python 3.9.1 (default, Dec 18 2020, 18:01:25) 
[GCC 10.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import ruamel.yaml
>>> dir(ruamel.yaml)
['AliasEvent', 'AliasToken', 'AnchorToken', 'BaseConstructor', 'BaseDumper', 'BaseLoader', 'BaseRepresenter', 'BlockEndToken', 'BlockEntryToken', 'BlockMappingStartToken', 'BlockSequenceStartToken', 'BytesIO', 'CBaseDumper', 'CBaseLoader', 'CDumper', 'CEmitter', 'CLoader', 'CParser', 'CSafeDumper', 'CSafeLoader', 'CollectionEndEvent', 'CollectionNode', 'CollectionStartEvent', 'CommentCheck', 'CommentToken', 'Constructor', 'DirectiveToken', 'DocumentEndEvent', 'DocumentEndToken', 'DocumentStartEvent', 'DocumentStartToken', 'Dumper', 'Event', 'FlowEntryToken', 'FlowMappingEndToken', 'FlowMappingStartToken', 'FlowSequenceEndToken', 'FlowSequenceStartToken', 'KeyToken', 'Loader', 'MappingEndEvent', 'MappingNode', 'MappingStartEvent', 'Node', 'NodeEvent', 'PY3', 'Representer', 'Resolver', 'RoundTripConstructor', 'RoundTripDumper', 'RoundTripLoader', 'RoundTripRepresenter', 'SHOWLINES', 'SafeConstructor', 'SafeDumper', 'SafeLoader', 'SafeRepresenter', 'ScalarEvent', 'ScalarNode', 'ScalarToken', 'SequenceEndEvent', 'SequenceNode', 'SequenceStartEvent', 'StreamEndEvent', 'StreamEndToken', 'StreamStartEvent', 'StreamStartToken', 'StringIO', 'TagToken', 'Token', 'UnsafeLoader', 'UnsafeLoaderWarning', 'ValueToken', 'VersionedResolver', 'YAML', 'YAMLContextManager', 'YAMLError', 'YAMLObject', 'YAMLObjectMetaclass', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', '__version__', '__with_libyaml__', '_package_data', 'absolute_import', 'add_constructor', 'add_implicit_resolver', 'add_multi_constructor', 'add_multi_representer', 'add_path_resolver', 'add_representer', 'anchor', 'comments', 'compat', 'compose', 'compose_all', 'composer', 'constructor', 'cyaml', 'division', 'dump', 'dump_all', 'dumper', 'emit', 'emitter', 'enc', 'enforce', 'error', 'events', 'glob', 'import_module', 'load', 'load_all', 'loader', 'main', 'nodes', 'nprint', 'os', 'parse', 'parser', 'pkg_resources', 'print_function', 'reader', 'representer', 'resolver', 'round_trip_dump', 'round_trip_load', 'round_trip_load_all', 'ruamel', 'safe_dump', 'safe_dump_all', 'safe_load', 'safe_load_all', 'scalarbool', 'scalarfloat', 'scalarint', 'scalarstring', 'scan', 'scanner', 'serialize', 'serialize_all', 'serializer', 'string_types', 'sys', 'timestamp', 'tokens', 'unicode_literals', 'util', 'version_info', 'warnings', 'with_metaclass', 'yaml_object']
>>> dir(ruamel.yaml.safe_load)
['__annotations__', '__call__', '__class__', '__closure__', '__code__', '__defaults__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__get__', '__getattribute__', '__globals__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__kwdefaults__', '__le__', '__lt__', '__module__', '__name__', '__ne__', '__new__', '__qualname__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__']

```

**Comment by asavah** [comment]
From my tests so far it seems to affect ruamel.yaml (0.6.12) other modules are ok.

**Comment by asavah** [comment]
Grepped tracebacks from python3 -v output

Good (51.1.2):
```
import 'pkg_resources' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d56656fa0>
import 'ruamel' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d56656df0>
import 'ruamel.yaml.compat' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d55e0ac70>
import 'ruamel.yaml.error' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d55e0a040>
import 'ruamel.yaml.util' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d55e11820>
import 'ruamel.yaml.reader' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d55e0aaf0>
import 'ruamel.yaml.tokens' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d55d36a60>
import 'ruamel.yaml.scanner' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d55e11c70>
import 'ruamel.yaml.events' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d55d47f40>
import 'ruamel.yaml.parser' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d55d36cd0>
import 'ruamel.yaml.nodes' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d55d57910>
import 'ruamel.yaml.composer' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d55d47d30>
import 'ruamel.yaml.anchor' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d55d103d0>
import 'ruamel.yaml.scalarstring' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d55d05b80>
import 'ruamel.yaml.comments' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d55d67df0>
import 'ruamel.yaml.scalarint' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d55cffaf0>
import 'ruamel.yaml.scalarfloat' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d55cfff10>
import 'ruamel.yaml.scalarbool' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d55d10d00>
import 'ruamel.yaml.timestamp' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d55d10fa0>
import 'ruamel.yaml.constructor' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d55d57b50>
import 'ruamel.yaml.emitter' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d55d57a90>
import 'ruamel.yaml.serializer' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d55d21f40>
import 'ruamel.yaml.representer' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d55cb4400>
import '_ruamel_yaml' # <_frozen_importlib_external.ExtensionFileLoader object at 0x7f2d55df5d90>
import 'ruamel.yaml.resolver' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d55e0a5e0>
import 'ruamel.yaml.cyaml' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d56523d90>
import 'glob' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d55ce12b0>
import 'ruamel.yaml.loader' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d55ce17c0>
import 'ruamel.yaml.dumper' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d55ce1a60>
import 'ruamel.yaml.main' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d5653e490>
import 'ruamel.yaml' # <_frozen_importlib_external.SourceFileLoader object at 0x7f2d56656d00>
```

Bad (51.3.3):
```
import 'pkg_resources' # <_frozen_importlib_external.SourceFileLoader object at 0x7f77c165cf40>
import 'ruamel' # <_frozen_importlib_external.SourceFileLoader object at 0x7f77c165cd90>
import 'ruamel.yaml' # <_frozen_importlib_external.SourceFileLoader object at 0x7f77c165cdc0>
```

**Comment by asavah** [comment]
Caused by https://github.com/pypa/setuptools/commit/f146b387548711d8095002e3d1214ca98a8b0392

**Comment by jaraco** [comment]
Appreciate you tracing the implicated change. That helps a lot.

Unfortunately, my findings don't match yours. For example, consider this command that builds the package from source using Setuptools 51.3.2, but the error still occurs:

```
$ pip-run -q setuptools==51.3.2 -- -m pip-run --no-build-isolation --no-binary ruamel.yaml ruamel.yaml -- -c "import ruamel.yaml.safe_load"
Collecting ruamel.yaml
  Using cached ruamel.yaml-0.16.12.tar.gz (147 kB)
Skipping wheel build for ruamel.yaml, due to binaries being disabled for it.
Installing collected packages: ruamel.yaml
    Running setup.py install for ruamel.yaml ... done
Successfully installed ruamel.yaml-0.16.12
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'ruamel.yaml.safe_load'
```

Oh, f146b38 isn't new in 51.3.3. It's new in 51.2.0. But even with the release prior to 51.2, the same error occurs:

```
setuptools main $ pip-run -q 'setuptools<51.2.0' -- -m pip-run --no-build-isolation --no-binary ruamel.yaml ruamel.yaml -- -c "import rua
mel.yaml.safe_load"
Collecting ruamel.yaml
  Using cached ruamel.yaml-0.16.12.tar.gz (147 kB)
Skipping wheel build for ruamel.yaml, due to binaries being disabled for it.
Installing collected packages: ruamel.yaml
    Running setup.py install for ruamel.yaml ... done
Successfully installed ruamel.yaml-0.16.12
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'ruamel.yaml.safe_load'
```

**Comment by jaraco** [comment]
I had a peek at [the source for ruamel.yaml](https://sourceforge.net/p/ruamel-yaml/code/ci/default/tree/setup.py), and I see it does a lot of nonstandard stuff (setup.py is almost 1000 lines).

I'm afraid I'm going to have to ask you to do some more analysis, to find more essentially what the issue is, what it is about this package that triggers this behavior, and if it is this change, what is it about the change that violates the expectations of ruamel.yaml?

**Comment by jaraco** [comment]
I did find I was able to resolve the `safe_load` attribute properly with late Setuptools:

```
$ pip-run -q 'setuptools==51.3.3' -- -m pip-run --no-build-isolation --no-binary ruamel.yaml ruamel.yaml -- -c "import ruamel.yaml;
 print(dir(ruamel.yaml.safe_load))"
Collecting ruamel.yaml
  Using cached ruamel.yaml-0.16.12.tar.gz (147 kB)
Skipping wheel build for ruamel.yaml, due to binaries being disabled for it.
Installing collected packages: ruamel.yaml
    Running setup.py install for ruamel.yaml ... done
Successfully installed ruamel.yaml-0.16.12
['__annotations__', '__call__', '__class__', '__closure__', '__code__', '__defaults__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__get__', '__getattribute__', '__globals__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__kwdefaults__', '__le__', '__lt__', '__module__', '__name__', '__ne__', '__new__', '__qualname__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__']
```

**Comment by asavah** [comment]
Thanks a lot for looking into this.
I just tested this with setuptools-52.0.0 and found that it fails only if ruamel.yaml.clib is present.  (namespace issue?)
ruamel.yaml requirements state ruamel.yaml.clib is needed for python <= 3.8 , I have 3.9 here.
Thus I'm dropping ruamel.yaml.clib for the time being.

It worked fine with python 3.9, clib and setuptools <= 51.3.3 though.

I'm aware that ruamel.yaml does a bunch of non standard things, sadly it's needed for the project I'm working on and rewriting
everything for another implementation eg pyyaml is not an option at the moment.

Thanks again for your time and feel free to close this issue.
I was not sure from whether this was a purely setuptools problem to begin with, but I decided to report it anyways because version bump triggered/revelead it.
